libdevcore/JsonUtils.cpp

```diff
@@ -31,3 +31,77 @@ void dev::validateFieldNames(json_spirit::mObject const& _obj, std::set<std::str
 			BOOST_THROW_EXCEPTION(UnknownField() << errinfo_comment(comment));
 		}
 }
+
+std::string dev::jsonTypeAsString(json_spirit::Value_type _type)
+{
+    switch (_type)
+    {
+    case json_spirit::obj_type:
+        return "json Object";
+    case json_spirit::array_type:
+        return "json Array";
+    case json_spirit::str_type:
+        return "json String";
+    case json_spirit::bool_type:
+        return "json Bool";
+    case json_spirit::int_type:
+        return "json Int";
+    case json_spirit::real_type:
+        return "json Real";
+    case json_spirit::null_type:
+        return "json Null";
+    default:
+        return "json n/a";
+    }
+}
+
+void dev::requireJsonFields(json_spirit::mObject const& _o, std::string const& _config,
+    std::map<std::string, possibleJsonType> const& _validationMap,
+    std::set<std::string> const& _ignoreFields)
+{
+    // check for unexpected fiedls
+    for (auto const& field : _o)
+    {
+        if (!_validationMap.count(field.first) && !_ignoreFields.count(field.first))
+        {
+            std::string const comment =
+                "Unexpected field '" + field.first + "' in config: " + _config;
+            std::cerr << comment << "\n"
+                      << json_spirit::write_string((json_spirit::mValue)_o, true) << "\n";
+            BOOST_THROW_EXCEPTION(UnknownField() << errinfo_comment(comment));
+        }
+    }
+
+    // check field types with validation map
+    for (auto const vmap : _validationMap)
+    {
+        if (!_o.count(vmap.first))
+        {
+            std::string const comment =
+                "Expected field '" + vmap.first + "' not found in config: " + _config;
+            std::cerr << comment << "\n"
+                      << json_spirit::write_string((json_spirit::mValue)_o, true) << "\n";
+            BOOST_THROW_EXCEPTION(MissingField() << errinfo_comment(comment));
+        }
+
+        bool matched = false;
+        std::string sTypes;
+        for (auto const& type : vmap.second)
+        {
+            if (sTypes.size())
+                sTypes += ", or ";
+            sTypes += jsonTypeAsString(type);
```

Comment:
You need this concatenation only when there is an error.

Reply:
so you think two loops on `for (auto const& type : vmap.second)` is better ?

Reply:
yes